File: 3154-maximum-value-of-an-ordered-triplet-i/3154-maximum-value-of-an-ordered-triplet-i.cs
public class Solution {
    public long MaximumTripletValue(int[] nums) {
        var n = nums.Length;
        long ans = 0, dmax = 0, imax = 0;
        foreach (var num in nums) {
            ans = Math.Max(ans, dmax * num);
            dmax = Math.Max(dmax, imax - num);
            imax = Math.Max(imax, num);
        }
        return ans;
    }
}

File: 3154-maximum-value-of-an-ordered-triplet-i/3154-maximum-value-of-an-ordered-triplet-i.dart
class Solution {
    int maximumTripletValue(List<int> nums) {
        int n = nums.length;
        int ans = 0, dmax = 0, imax = 0;
        for (int i = 0; i < n; ++i) {
            if (dmax * nums[i] > ans)
                ans = dmax * nums[i];
            if (imax - nums[i] > dmax)
                dmax = imax - nums[i];
            if (nums[i] > imax)
                imax = nums[i];
        }
        return ans;
    }
}